### ISSUE TITLE
JS compiler cannot compile some .ksy files

### ISSUE THREAD

**Issue by koczkatamas** [kaitai-io/kaitai_struct_compiler#23]
I found the following problems while I tried to compile formats from the kaitai_struct_formats repo with the JS compiler:
- `scala.scalajs.runtime.UndefinedBehaviorError: An undefined behavior was detected: [object Object] is not an instance of java.util.Map`
  - standard_midi_file.ksy
- `java.lang.RuntimeException: Unable to parse fixed content: null`
  - jpeg.ksy
  - png.ksy
  - wmf.ksy
  - icmp_packet.ksy
  - pcap.ksy
  - ext2.ksy
  - iso9660.ksy
  - mbr_partition_table.ksy
  - java_class.ksy
  - microsoft_pe.ksy

We should run test automatically which compiles every format descriptor in the repo with the JS compiler too.

**Comment by GreyCat** [comment]
Diagnosed the problems, both are related to the way we parse YAML files:

1. The problem arises in [AttrSpec.parseSwitch](https://github.com/kaitai-io/kaitai_struct_compiler/blob/master/shared/src/main/scala/io/kaitai/struct/format/AttrSpec.scala#L155), when parsing any switch type clause. This happens because we rely on accessing Java's `Map` in a code that is shared between JVM and JS implementations.

2. The problem is in [AttrSpec.parseContentSpec](https://github.com/kaitai-io/kaitai_struct_compiler/blob/master/shared/src/main/scala/io/kaitai/struct/format/AttrSpec.scala#L50), which tries to parse internals of content spec using Java classes like `ArrayList`, which do not happen to be used when data's coming from JS.

The universal solution for both would be having a "proper" YAML parser written in Scala, same for JVM and JS, so we won't have to rely on Java classes and/or JavaScript libraries here. In addition, it would allow better error messages, closer control on the parsing, etc. However, this is pretty difficult and long-term goal.

In short term, I believe first problem can be solved by adding extra class that would represent the switch type and do the parsing with two implemntations in both JS and JVM. I'm not sure about the second problem, as it turns to be fairly hard to do "proper" `@JsonCreator` / `@JsonProperty` mapping in this case, as they don't generally allow any simple way to parse polymorphic structures, i.e. which can be both a string or a map.

**Comment by GreyCat** [comment]
I've checked that both problems are solved when using the code proposed in https://github.com/kaitai-io/kaitai_struct/issues/42

**Comment by GreyCat** [comment]
Unified YAML parsing merged to master, so I consider these problems to be fixed.
